metaflow/plugins/kfp/tests/flows/resources_flow.py

```diff
@@ -0,0 +1,90 @@
+import os
+import pprint
+from typing import Dict, List
+
+from kubernetes.client import (
+    V1EnvVar,
+    V1EnvVarSource,
+    V1ObjectFieldSelector,
+    V1ResourceFieldSelector,
+)
+
+from metaflow import FlowSpec, step, environment, resources
+
+
+def get_env_vars(env_resources: Dict[str, str]) -> List[V1EnvVar]:
+    res = []
+    for name, resource in env_resources.items():
+        res.append(
+            V1EnvVar(
+                # this is used by some functions of operator-sdk
+                # it uses this environment variable to get the pods
+                name=name,
+                value_from=V1EnvVarSource(
+                    resource_field_ref=V1ResourceFieldSelector(
+                        container_name="main",
+                        resource=resource,
+                        divisor="1m" if "cpu" in resource else "1",
+                    )
+                ),
+            )
+        )
+    return res
+
+
+kubernetes_vars = get_env_vars(
+    {
+        "LOCAL_STORAGE": "requests.ephemeral-storage",
+        "LOCAL_STORAGE_LIMIT": "limits.ephemeral-storage",
+        "CPU": "requests.cpu",
+        "CPU_LIMIT": "limits.cpu",
+        "MEMORY": "requests.memory",
+        "MEMORY_LIMIT": "limits.memory",
+    }
+)
+kubernetes_vars.append(
+    V1EnvVar(
+        name="POD_NAME",
+        value_from=V1EnvVarSource(
+            field_ref=V1ObjectFieldSelector(field_path="metadata.name")
+        ),
+    )
+)
+
+
+class ResourcesFlow(FlowSpec):
+    @resources(
+        local_storage="100",
+        local_storage_limit="242",
+        cpu="0.1",
+        cpu_limit="0.6",
+        memory="500",
+        memory_limit="1G",
+    )
+    @environment(vars={"MY_ENV": "value"}, kubernetes_vars=kubernetes_vars)
```

Comment:
Just clarifying since the way these kubernetes_vars getting into os.environ seems implicit to me.

The normal `vars` uses the following but I don't see anything similar for `kubernetes_vars`
```
def step_init(self, flow, graph, step, decos, environment, datastore, logger):
    os.environ.update(self.attributes['vars'].items())
```

Are they compiled into argo file in a format recognizable by Kubernetes, and rendered by Kubernetes at deploy time?

I appreciate `kubernetes_vars` as a very good unit testing short cut, but if we want to make this available to customer as a feature we might want to encapsulate `get_env_vars` function inside workflow SDK and put more documentation.

Reply:
I just saw your conversation with Alex above. I think you answered how this value gets to Metaflow `Decorator.attributes`, but I am mainly curious about how it gets into environment variables.

Reply:
How they get into environment vars is part of this PR:
https://github.com/zillow/metaflow/blob/cc1e5318f610c77b4cee237dc037510c35d292cc/metaflow/plugins/kfp/kfp.py#L715-L721

Reply:
The decorator doesn't have access to the KFP `container_op` for the the step.  This is the same when setting container resources for the `@resources` decorator https://github.com/zillow/metaflow/blob/cc1e5318f610c77b4cee237dc037510c35d292cc/metaflow/plugins/kfp/kfp.py#L752-L752

Reply:
I see the complication... Are we able to add container_op. at compile time in later PRs? Ideally we want
```
@environment(vars={"MY_ENV": "value"}, kubernetes_vars={
    "LOCAL_STORAGE": "requests.ephemeral-storage",
    "LOCAL_STORAGE_LIMIT": "limits.ephemeral-storage",
    "CPU": "requests.cpu",
    "CPU_LIMIT": "limits.cpu",
    "MEMORY": "requests.memory",
    "MEMORY_LIMIT": "limits.memory",
})
```
and do this in the line you just [linked](https://github.com/zillow/metaflow/blob/cc1e5318f610c77b4cee237dc037510c35d292cc/metaflow/plugins/kfp/kfp.py#L715-L721) : 
```
for name, resource in envs.items(): 
    container_op.container.add_env_variable(
        V1EnvVar(
            name=name,
            value_from=V1EnvVarSource(
                resource_field_ref=V1ResourceFieldSelector(
                    container_name="main",
                    resource=resource,
                    divisor="1m" if "cpu" in resource else "1",
                )
            )
        )
    ) 
```
Never mind if this doesn't work... I haven't tried this myself and may mis-thinking it.